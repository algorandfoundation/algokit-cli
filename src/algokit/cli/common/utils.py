# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
