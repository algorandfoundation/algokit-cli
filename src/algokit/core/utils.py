import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
