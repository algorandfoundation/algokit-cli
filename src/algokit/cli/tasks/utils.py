# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
