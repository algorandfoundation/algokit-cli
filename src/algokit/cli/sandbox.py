import logging

import click

from algokit.cli.explore import explore_command
from algokit.cli.goal import goal_command
from algokit.core import proc
from algokit.core.sandbox import (
    DOCKER_COMPOSE_MINIMUM_VERSION,
    DOCKER_COMPOSE_VERSION_COMMAND,
    ComposeFileStatus,
    ComposeSandbox,
    fetch_algod_status_data,
    fetch_indexer_status_data,
    parse_docker_compose_version_output,
)
from algokit.core.utils import is_minimum_version

logger = logging.getLogger(__name__)


@click.group("sandbox", short_help="Manage the AlgoKit sandbox.")
def sandbox_group() -> None:
    try:
        compose_version_result = proc.run(DOCKER_COMPOSE_VERSION_COMMAND)
    except IOError as ex:
        # an IOError (such as PermissionError or FileNotFoundError) will only occur if "docker"
        # isn't an executable in the user's path, which means docker isn't installed
        raise click.ClickException(
            "Docker not found; please install Docker and add to path.\n"
            "See https://docs.docker.com/get-docker/ for more information."
        ) from ex
    if compose_version_result.exit_code != 0:
        raise click.ClickException(
            "Docker Compose not found; please install Docker Compose and add to path.\n"
            "See https://docs.docker.com/compose/install/ for more information."
        )
    compose_version_str = parse_docker_compose_version_output(compose_version_result.output)
    try:
        compose_version_ok = is_minimum_version(compose_version_str, DOCKER_COMPOSE_MINIMUM_VERSION)
    except Exception:
        logger.warning(
            "Unable to extract docker compose version from output: \n"
            + compose_version_str
            + f"\nPlease ensure a minimum of compose v{DOCKER_COMPOSE_MINIMUM_VERSION} is used",
            exc_info=True,
        )
    else:
        if not compose_version_ok:
            raise click.ClickException(
                f"Minimum docker compose version supported: v{DOCKER_COMPOSE_MINIMUM_VERSION}, "
                f"installed = v{compose_version_str}\n"
                "Please update your Docker install"
            )

    proc.run(["docker", "version"], bad_return_code_error_message="Docker engine isn't running; please start it.")


@sandbox_group.command("start", short_help="Start the AlgoKit Sandbox.")
def start_sandbox() -> None:
    sandbox = ComposeSandbox()
    compose_file_status = sandbox.compose_file_status()
    if compose_file_status is ComposeFileStatus.MISSING:
        logger.debug("Sandbox compose file does not exist yet; writing it out for the first time")
        sandbox.write_compose_file()
    elif compose_file_status is ComposeFileStatus.UP_TO_DATE:
        logger.debug("Sandbox compose file does not require updating")
    else:
        logger.warning("Sandbox definition is out of date; please run algokit sandbox reset")
    sandbox.up()


@sandbox_group.command("stop", short_help="Stop the AlgoKit Sandbox.")
def stop_sandbox() -> None:
    sandbox = ComposeSandbox()
    compose_file_status = sandbox.compose_file_status()
    if compose_file_status is ComposeFileStatus.MISSING:
        logger.debug("Sandbox compose file does not exist yet; run `algokit sandbox start` to start the Sandbox")
    else:
        sandbox.stop()


@sandbox_group.command("reset", short_help="Reset the AlgoKit Sandbox.")
@click.option(
    "--update/--no-update",
    default=True,
    help="Enable or disable updating to the latest available Sandbox version",
)
def reset_sandbox(update: bool) -> None:  # noqa: FBT001
    sandbox = ComposeSandbox()
    compose_file_status = sandbox.compose_file_status()
    if compose_file_status is ComposeFileStatus.MISSING:
        logger.debug("Existing Sandbox not found; creating from scratch...")
        sandbox.write_compose_file()
    else:
        sandbox.down()
        if compose_file_status is not ComposeFileStatus.UP_TO_DATE:
            logger.info("Sandbox definition is out of date; updating it to latest")
            sandbox.write_compose_file()
        if update:
            sandbox.pull()
    sandbox.up()


SERVICE_NAMES = ("algod", "indexer", "indexer-db")


@sandbox_group.command("status", short_help="Check the status of the AlgoKit Sandbox.")
def sandbox_status() -> None:
    sandbox = ComposeSandbox()
    ps = sandbox.ps()
    ps_by_name = {stats["Service"]: stats for stats in ps}
    # if any of the required containers does not exist (ie it's not just stopped but hasn't even been created),
    # then they will be missing from the output dictionary
    if set(SERVICE_NAMES) != ps_by_name.keys():
        raise click.ClickException("Sandbox has not been initialized yet, please run 'algokit sandbox start'")
    # initialise output dict by setting status
    output_by_name = {
        name: {"Status": "Running" if ps_by_name[name]["State"] == "running" else "Not running"}
        for name in SERVICE_NAMES
    }
    # fill out remaining output_by_name["algod"] values
    if output_by_name["algod"]["Status"] == "Running":
        output_by_name["algod"].update(fetch_algod_status_data(ps_by_name["algod"]))
    # fill out remaining output_by_name["indexer"] values
    if output_by_name["indexer"]["Status"] == "Running":
        output_by_name["indexer"].update(fetch_indexer_status_data(ps_by_name["indexer"]))

    # Print the status details
    for service_name, service_info in output_by_name.items():
        logger.info(click.style(f"# {service_name} status", bold=True))
        for key, value in service_info.items():
            logger.info(click.style(f"{key}:", bold=True) + f" {value}")

    # return non-zero if any container is not running
    if not all(item["Status"] == "Running" for item in output_by_name.values()):
        raise click.ClickException(
            "At least one container isn't running; execute `algokit sandbox start` to start the Sandbox"
        )


@sandbox_group.command(
    "console",
    short_help="Run the Algorand goal CLI against the AlgoKit Sandbox via a Bash console"
    + " so you can execute multiple goal commands and/or interact with a filesystem.",
)
@click.pass_context
def sandbox_console(context: click.Context) -> None:
    context.invoke(goal_command, console=True)


@sandbox_group.command("explore", short_help="Explore the AlgoKit Sandbox using Dappflow")
@click.pass_context
def sandbox_explore(context: click.Context) -> None:
    context.invoke(explore_command)
