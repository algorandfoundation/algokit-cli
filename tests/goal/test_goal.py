import json
from pathlib import Path
from subprocess import CompletedProcess

import pytest
from algokit.core.sandbox import get_config_json, get_docker_compose_yml
from pytest_mock import MockerFixture

from tests.utils.app_dir_mock import AppDirs
from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock

DUMMY_CONTRACT_TEAL = """\n#pragma version 8\nint 1\nreturn\n"""


def _normalize_output(output: str) -> str:
    return output.replace("\\", "/")


@pytest.fixture()
def cwd(tmp_path_factory: pytest.TempPathFactory) -> Path:
    return tmp_path_factory.mktemp("cwd")


@pytest.fixture()
def mocked_goal_mount_path(cwd: Path, monkeypatch: pytest.MonkeyPatch) -> Path:
    mocked_goal_mount = cwd / "goal_mount"
    mocked_goal_mount.mkdir()
    monkeypatch.setattr("algokit.cli.goal.get_volume_mount_path_local", lambda: cwd / "goal_mount")
    return mocked_goal_mount


@pytest.fixture()
def _setup_latest_dummy_compose(app_dir_mock: AppDirs) -> None:
    (app_dir_mock.app_config_dir / "sandbox").mkdir()
    (app_dir_mock.app_config_dir / "sandbox" / "docker-compose.yml").write_text(get_docker_compose_yml())
    (app_dir_mock.app_config_dir / "sandbox" / "algod_config.json").write_text(get_config_json())


@pytest.fixture()
def _setup_input_files(cwd: Path, request: pytest.FixtureRequest) -> None:
    files = request.param
    for file in files:
        if "name" in file:
            if "content" in file:
                (cwd / file["name"]).write_text(file["content"], encoding="utf-8")
            else:
                (cwd / file["name"]).touch()

            assert (cwd / file["name"]).exists()


def dump_file(cwd: Path) -> None:
    (cwd / "approval.compiled").write_text(
        """
I AM COMPILED!
""",
        encoding="utf-8",
    )


def dump_json_file(cwd: Path) -> None:
    (cwd / "balance_record.json").write_text(
        """
I AM COMPILED!
""",
        encoding="utf-8",
    )


def test_goal_help() -> None:
    result = invoke("goal -h")

    assert result.exit_code == 0
    verify(result.output)


@pytest.mark.usefixtures("proc_mock", "_setup_latest_dummy_compose", "mocked_goal_mount_path")
def test_goal_no_args() -> None:
    result = invoke("goal")

    assert result.exit_code == 0
    verify(result.output)


@pytest.mark.usefixtures("proc_mock", "_setup_latest_dummy_compose")
def test_goal_console(mocker: MockerFixture) -> None:
    mocker.patch("algokit.core.proc.subprocess_run").return_value = CompletedProcess(
        ["docker", "exec"], 0, "STDOUT+STDERR"
    )

    result = invoke("goal --console")

    assert result.exit_code == 0
    verify(result.output)


@pytest.mark.usefixtures("_setup_latest_dummy_compose")
def test_goal_console_failed(app_dir_mock: AppDirs, proc_mock: ProcMock, mocker: MockerFixture) -> None:
    mocker.patch("algokit.core.proc.subprocess_run").return_value = CompletedProcess(
        ["docker", "exec"], 1, "STDOUT+STDERR"
    )

    proc_mock.set_output(
        ["docker", "compose", "ps", "algod", "--format", "json"],
        output=[json.dumps([{"Name": "algokit_algod", "State": "running"}])],
    )

    result = invoke("goal --console")

    assert result.exit_code == 1
    verify(_normalize_output(result.output.replace(str(app_dir_mock.app_config_dir), "{app_config}")))


@pytest.mark.usefixtures("_setup_latest_dummy_compose")
def test_goal_console_failed_algod_not_created(
    app_dir_mock: AppDirs, proc_mock: ProcMock, mocker: MockerFixture
) -> None:
    mocker.patch("algokit.core.proc.subprocess_run").return_value = CompletedProcess(
        ["docker", "exec"], 1, "bad args to goal"
    )

    proc_mock.set_output(["docker", "compose", "ps", "algod", "--format", "json"], output=[json.dumps([])])

    result = invoke("goal --console")

    assert result.exit_code == 1
    verify(_normalize_output(result.output.replace(str(app_dir_mock.app_config_dir), "{app_config}")))


@pytest.mark.usefixtures("proc_mock", "_setup_latest_dummy_compose", "mocked_goal_mount_path")
def test_goal_simple_args() -> None:
    result = invoke("goal account list")

    assert result.exit_code == 0
    verify(result.output)


@pytest.mark.usefixtures("proc_mock", "_setup_latest_dummy_compose", "mocked_goal_mount_path")
def test_goal_complex_args() -> None:
    result = invoke("goal account export -a RKTAZY2ZLKUJBHDVVA3KKHEDK7PRVGIGOZAUUIZBNK2OEP6KQGEXKKUYUY")

    assert result.exit_code == 0
    verify(result.output)


def test_goal_start_without_docker(proc_mock: ProcMock) -> None:
    proc_mock.should_fail_on("docker version")

    result = invoke("goal")

    assert result.exit_code == 1
    verify(result.output)


def test_goal_start_without_docker_engine_running(proc_mock: ProcMock) -> None:
    proc_mock.should_bad_exit_on("docker version")

    result = invoke("goal")

    assert result.exit_code == 1
    verify(result.output)


@pytest.mark.usefixtures("_setup_input_files", "_setup_latest_dummy_compose", "mocked_goal_mount_path")
@pytest.mark.parametrize("_setup_input_files", [[{"name": "transactions.txt"}]], indirect=True)
def test_goal_simple_args_with_input_file(
    proc_mock: ProcMock,
    cwd: Path,
) -> None:
    expected_arguments = [
        "docker",
        "exec",
        "--interactive",
        "--workdir",
        "/root",
        "algokit_algod",
        "goal",
        "clerk",
        "group",
    ]

    proc_mock.set_output(expected_arguments, output=["File compiled"])
    result = invoke("goal clerk group transactions.txt", cwd=cwd)

    # Check if the path in command has changed in preprocess step
    assert _normalize_output(proc_mock.called[1].command[9]) == "/root/goal_mount/transactions.txt"

    # Check for the result status
    assert result.exit_code == 0

    verify(_normalize_output(result.output))


@pytest.mark.usefixtures("mocked_goal_mount_path", "_setup_latest_dummy_compose")
def test_goal_simple_args_with_output_file(proc_mock: ProcMock, cwd: Path) -> None:
    expected_arguments = [
        "docker",
        "exec",
        "--interactive",
        "--workdir",
        "/root",
        "algokit_algod",
        "goal",
        "account",
        "dump",
    ]

    proc_mock.set_output(
        expected_arguments,
        output=["File compiled"],
        side_effect=dump_json_file,
        side_effect_args={"cwd": cwd},
    )
    result = invoke("goal account dump -o balance_record.json")

    # Check if the path in command has changed in preprocess step
    assert _normalize_output(proc_mock.called[1].command[10]) == "/root/goal_mount/balance_record.json"

    # Check for the result status
    assert result.exit_code == 0

    # Check if the output file is actually created and copied in cwd in postprocess step
    assert (cwd / "balance_record.json").exists()

    verify(_normalize_output(result.output))


@pytest.mark.usefixtures("mocked_goal_mount_path", "_setup_input_files", "_setup_latest_dummy_compose")
@pytest.mark.parametrize(
    "_setup_input_files", [[{"name": "approval.teal", "content": DUMMY_CONTRACT_TEAL}]], indirect=True
)
def test_goal_simple_args_with_input_output_files(
    proc_mock: ProcMock,
    cwd: Path,
) -> None:
    expected_arguments = [
        "docker",
        "exec",
        "--interactive",
        "--workdir",
        "/root",
        "algokit_algod",
        "goal",
        "clerk",
        "compile",
    ]

    proc_mock.set_output(
        expected_arguments, output=["File compiled"], side_effect=dump_file, side_effect_args={"cwd": cwd}
    )

    result = invoke("goal clerk compile approval.teal -o approval.compiled", cwd=cwd)

    # Check if the paths in command have changed in preprocess step
    assert _normalize_output(proc_mock.called[1].command[9]) == "/root/goal_mount/approval.teal"
    assert _normalize_output(proc_mock.called[1].command[11]) == "/root/goal_mount/approval.compiled"

    # Check for the result status
    assert result.exit_code == 0

    # Check if the output file is created and copied in cwd in postprocess step
    assert (cwd / "approval.compiled").exists()
    verify(_normalize_output(result.output))


@pytest.mark.usefixtures("mocked_goal_mount_path", "_setup_input_files", "_setup_latest_dummy_compose")
@pytest.mark.parametrize(
    "_setup_input_files",
    [
        [
            {"name": "approval1.teal", "content": DUMMY_CONTRACT_TEAL},
            {"name": "approval2.teal", "content": DUMMY_CONTRACT_TEAL},
        ]
    ],
    indirect=True,
)
def test_goal_simple_args_with_multiple_input_output_files(
    proc_mock: ProcMock,
    cwd: Path,
) -> None:
    expected_arguments = [
        "docker",
        "exec",
        "--interactive",
        "--workdir",
        "/root",
        "algokit_algod",
        "goal",
        "clerk",
        "compile",
    ]

    proc_mock.set_output(
        expected_arguments, output=["File compiled"], side_effect=dump_file, side_effect_args={"cwd": cwd}
    )
    result = invoke("goal clerk compile approval1.teal approval2.teal -o approval.compiled", cwd=cwd)

    # Check if the paths in command have changed in preprocess step
    assert _normalize_output(proc_mock.called[1].command[9]) == "/root/goal_mount/approval1.teal"
    assert _normalize_output(proc_mock.called[1].command[10]) == "/root/goal_mount/approval2.teal"
    assert _normalize_output(proc_mock.called[1].command[12]) == "/root/goal_mount/approval.compiled"

    # Check for the result
    assert result.exit_code == 0

    # Check if the output file is actually created and copied in cwd in postprocess step
    assert (cwd / "approval.compiled").exists()
    verify(_normalize_output(result.output))


@pytest.mark.usefixtures("proc_mock", "mocked_goal_mount_path", "_setup_latest_dummy_compose")
def test_goal_simple_args_without_file_error(
    cwd: Path,
) -> None:
    assert not (cwd / "approval.teal").exists()
    result = invoke("goal clerk compile approval.teal -o approval.compiled", cwd=cwd)

    assert result.exit_code == 1
    verify(_normalize_output(result.output))


@pytest.mark.usefixtures("_setup_input_files", "_setup_latest_dummy_compose")
@pytest.mark.parametrize(
    "_setup_input_files", [[{"name": "approval.teal", "content": DUMMY_CONTRACT_TEAL}]], indirect=True
)
def test_goal_postprocess_of_command_args(
    proc_mock: ProcMock,
    cwd: Path,
    mocked_goal_mount_path: Path,
) -> None:
    # adding some dummy files to the mocked_goal_mount_path
    (mocked_goal_mount_path / "approval.group").touch()
    (mocked_goal_mount_path / "approval.group.sig").touch()
    (mocked_goal_mount_path / "approval.group.sig.out").touch()

    expected_arguments = [
        "docker",
        "exec",
        "--interactive",
        "--workdir",
        "/root",
        "algokit_algod",
        "goal",
        "clerk",
        "compile",
    ]
    proc_mock.set_output(
        expected_arguments,
        output=["File compiled"],
        side_effect=dump_file,
        side_effect_args={"cwd": mocked_goal_mount_path},
    )

    result = invoke("goal clerk compile approval.teal -o approval.compiled", cwd=cwd)
    assert result.exit_code == 0

    # check if the output files is no longer in the goal_mount_path
    assert not (mocked_goal_mount_path / "approval.compiled").exists()

    # check if the input/output file is in the cwd
    assert (cwd / "approval.compiled").exists()
    assert (cwd / "approval.teal").exists()

    # check if the dummy files are still there
    assert (mocked_goal_mount_path / "approval.group").exists()
    assert (mocked_goal_mount_path / "approval.group.sig").exists()
    assert (mocked_goal_mount_path / "approval.group.sig.out").exists()


@pytest.mark.usefixtures("proc_mock")
def test_goal_compose_outdated(
    cwd: Path,
    app_dir_mock: AppDirs,
) -> None:
    (app_dir_mock.app_config_dir / "sandbox").mkdir()
    (app_dir_mock.app_config_dir / "sandbox" / "docker-compose.yml").write_text("Outdated")
    (app_dir_mock.app_config_dir / "sandbox" / "algod_config.json").write_text("Outdated")

    result = invoke("goal help", cwd=cwd)

    assert result.exit_code == 1

    verify(_normalize_output(result.output))
