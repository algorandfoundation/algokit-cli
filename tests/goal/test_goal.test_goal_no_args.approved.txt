DEBUG: Running 'docker version' in '{current_working_directory}'
DEBUG: docker: STDOUT
DEBUG: docker: STDERR
DEBUG: Running 'docker exec --interactive --workdir /root algokit_algod goal' in '{current_working_directory}'
 STDOUT
 STDERR
